test: [DI-21337] - Add Cypress tests to verify ACLP UI's handling of API errors

agorthi-akamai opened the pull request:
Add Cypress tests to verify ACLP UI's handling of API errors

## Description 📝
This pull request focuses on validating the error-handling mechanisms for various API endpoints in the system. The following use cases are being validated to ensure appropriate error messages are displayed when API failures occur:

## Changes  🔄
As part of this PR, we are validating the following use cases:

Handling error for the Service Types API with the message: "Failed to fetch service types."
Handling error for the Dashboards API with the message: "Failed to fetch dashboards for Linode, DBaaS, etc."
Handling error for the Regions API with the message: "Failed to fetch regions."
Handling error for the Resources/Clusters API with the message: "Failed to fetch resources" or "Failed to fetch database clusters."
Handling error for the Dashboard by ID API with the message: "Failed to fetch the selected dashboard."

Jira ticket : https://track.akamai.com/jira/browse/DI-21337

## Target release date 🗓️ 
12/11/2024

## How to test 🧪
 yarn cy:run -s "cypress/e2e/core/cloudpulse/cloudpulse-validation.spec.ts"

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

jaalah-akamai commented:
@linode/frontend-sdet Please review the previously closed PR for historical context of comments made

jdamore-linode commented on packages/manager/cypress/e2e/core/cloudpulse/cloudpulse-dashboard-errors.spec.ts:
```diff
@@ -0,0 +1,396 @@
+const databaseMock: Database = databaseFactory.build({
+  label: clusterName,
+  type: engine,
+  region: region,
```
```suggestion
  region: mockRegion.id,
```
This value should be an API region ID, like `us-ord`.

jdamore-linode commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -139,3 +140,167 @@ export const mockCreateCloudPulseJWEToken = (
+ * This function intercepts the `GET` request made to the CloudPulse API and
+ * simulates an error response with a customizable error message and HTTP status code.
+ *
+ * @param {string} errorMessage - The error message to include in the mock response body.
+ * @param {string} serviceType - The service type for which the metric definitions are being mocked.
+ * @param {number} [status=500] - The HTTP status code for the mock response (defaults to 500 if not provided).
+ *
+ * @returns {Cypress.Chainable<null>} - A Cypress chainable object, indicating that the interception is part of a Cypress test chain.
+ */
+export const mockGetCloudPulseMetricDefinitionError = (
+  errorMessage: string,
+  serviceType: string,
```
```suggestion
 * @param {string} serviceType - The service type for which the metric definitions are being mocked.
 * @param {string} errorMessage - The error message to include in the mock response body.
 * @param {number} [status=500] - The HTTP status code for the mock response (defaults to 500 if not provided).
 *
 * @returns {Cypress.Chainable<null>} - A Cypress chainable object, indicating that the interception is part of a Cypress test chain.
 */
export const mockGetCloudPulseMetricDefinitionError = (
  serviceType: string,
  errorMessage: string,
```
Params that define what requests gets intercepted should come first, and params that define what gets mocked in response should come after. ([Previous discussion](https://github.com/linode/manager/pull/10891#discussion_r1784821070))

jdamore-linode commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -139,3 +140,167 @@ export const mockCreateCloudPulseJWEToken = (
+ * @returns {Cypress.Chainable<null>} - A Cypress chainable object, indicating that the interception is part of a Cypress test chain.
+ */
+export const mockGetCloudPulseTokenError = (
+  errorMessage: string,
+  serviceType: string,
```
Same as above

jdamore-linode commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -139,3 +140,167 @@ export const mockCreateCloudPulseJWEToken = (
+ * @returns {Cypress.Chainable<null>} - A Cypress chainable object, indicating that the interception is part of a Cypress test chain.
+ */
+export const mockGetCloudPulseDashboardsError = (
+  errorMessage: string,
+  serviceType: string,
```
Same as above

jdamore-linode commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -139,3 +140,167 @@ export const mockCreateCloudPulseJWEToken = (
+ * @returns {Cypress.Chainable<null>} - A Cypress chainable object, indicating that the interception is part of a Cypress test chain.
+ */
+export const mockGetCloudPulseDashboardByIdError = (
+  errorMessage: string,
+  id: number,
```
Same as above

jdamore-linode commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -139,3 +140,167 @@ export const mockCreateCloudPulseJWEToken = (
+ *
+ * @returns {Cypress.Chainable<null>} - A Cypress chainable object, indicating that the interception is part of a Cypress test chain.
+ */
+export const mockGetCloudPulseMetricDefinitionError = (
```
```suggestion
export const mockGetCloudPulseMetricDefinitionsError = (
```
nit: just making this naming consistent with the existing `mockGetCloudPulseMetricDefinitions` util

jdamore-linode commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -139,3 +140,167 @@ export const mockCreateCloudPulseJWEToken = (
+  );
+};
+
+/**
+ * Mocks an error response for the GET request to retrieve database instances in CloudPulse.
+ *
+ * This function intercepts the 'GET' request made to the CloudPulse API endpoint for retrieving database instances
+ * and simulates an error response with a customizable error message and HTTP status code.
+ *
+ * @param {string} errorMessage - The error message to include in the mock response body.
+ * @param {number} [status=500] - The HTTP status code for the mock response (defaults to 500 if not provided).
+ *
+ * @returns {Cypress.Chainable<null>} - A Cypress chainable object, indicating that the interception is part of a Cypress test chain.
+ */
+export const mockGetCloudPulseDatabaseInstancesError = (
+  errorMessage: string,
+  status: number = 500
+): Cypress.Chainable<null> => {
+  return cy.intercept(
+    'GET',
+    apiMatcher('databases/instances*'),
+    makeErrorResponse(errorMessage, status)
+  );
+};
```
This one doesn't actually belong here since this is a DBaaS endpoint, not a CloudPulse endpoint.

Instead, this should be defined in `support/intercepts/databases.ts` and should be named `mockGetDatabasesError`. Feel free to take a quick look at the [code for `mockGetDatabases`](https://github.com/linode/manager/blob/develop/packages/manager/cypress/support/intercepts/databases.ts#L47), but your implementation here looks good aside from the name.

jdamore-linode commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -139,3 +140,167 @@ export const mockCreateCloudPulseJWEToken = (
+ *
+ * @returns {Cypress.Chainable<null>} - A Cypress chainable object, indicating that the interception is part of a Cypress test chain.
+ */
+export const mockGetCloudPulseRegionsError = (
+  errorMessage: string,
+  status: number = 500
+): Cypress.Chainable<null> => {
+  return cy.intercept(
+    'GET',
+    apiMatcher('regions*'),
+    makeErrorResponse(errorMessage, status)
+  );
+};
```
Same issue as above: this belongs in `support/intercepts/regions.ts` and should be named `mockGetRegionsError`, but otherwise this looks good.

jdamore-linode reviewed (commented):
Thanks @agorthi-akamai, looking pretty good! Couple nitpicks and suggestions for your intercepts but I think this is pretty close to being ready

agorthi-akamai commented:
I've addressed all the feedback provided in the previous code review. All comments and suggestions have been incorporated. Please let me know if you have any further concerns or suggestions.

agorthi-akamai commented:
> Thanks @agorthi-akamai, looking pretty good! Couple nitpicks and suggestions for your intercepts but I think this is pretty close to being ready

Thanks for the feedback, @jdamore-linode ! I've addressed all the nitpicks and suggestions regarding the intercepts, and everything should be good to go now. Please let me know if you have any further questions or if there's anything else I can do.

jdamore-linode commented on packages/manager/.changeset/pr-11239-tests-1731305530868.md:
```diff
@@ -0,0 +1,5 @@
+"@linode/manager": Tests
+---
+
+Added Cypress tests to verify ACLP UI's handling of API ([#11239](https://github.com/linode/manager/pull/11239))
```
```suggestion
Add Cypress tests to verify ACLP UI's handling of API errors ([#11239](https://github.com/linode/manager/pull/11239))
```
Another small nit, sorry I missed this yesterday!
(Related docs: https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset, first bullet point)

agorthi-akamai replied:
Thanks, I have updated the PR and improved the line

jdamore-linode reviewed (approved):
Looks good, thanks @agorthi-akamai!

kmuddapo commented:
@jaalah-akamai   Could you please approve this and merge.  Please note these changes have been already approved by you on the other PR which got closed :). Thanks!!

pmakode-akamai reviewed (approved):
thanks @agorthi-akamai! 

All checks pass ✅

![Screenshot 2024-11-15 at 8 27 29 PM](https://github.com/user-attachments/assets/40e2d114-2e54-4f62-bbc1-192e6fd39fd2)

Could we update the changeset file as suggested by @jdamore-linode?
